tests/unit/test_register_user_action.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Register user action unit tests."""
+
+# pylint: disable=protected-access
+
+import typing
+import unittest.mock
+
+import ops
+import pytest
+from ops.charm import ActionEvent
+from ops.testing import Harness
+
+import synapse
+from user import User
+
+
+@pytest.mark.parametrize("harness", [0], indirect=True)
+def test_register_user_action(
+    harness_server_name_configured: Harness, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """
+    arrange: start the Synapse charm, set Synapse container to be ready and set server_name.
+    act: run register-user action.
+    assert: Synapse charm should reset the instance.
+    """
+    harness = harness_server_name_configured
+    harness.set_leader(True)
+    get_registration_mock = unittest.mock.Mock(return_value="shared_secret")
+    monkeypatch.setattr("synapse.get_registration_shared_secret", get_registration_mock)
+    register_user_mock = unittest.mock.MagicMock()
+    monkeypatch.setattr("synapse.register_user", register_user_mock)
+    user = "username"
+    event = unittest.mock.MagicMock(spec=ActionEvent)
+    event.params = {
+        "username": user,
+        "admin": "no",
+    }
+    # Calling to test the action since is not possible calling via harness
```

Comment:
@benhoyt is that right? The harness can't trigger actions?

Reply:
I believe you can trigger an action more directly with `emit`, but unfortunately the Harness currently stubs out the testing `action_set` (and similar functions) used by `event.set_results`, hence why this needs to be mocked.

Yeah, we should provide Harness support for this. Feel free to open an issue to track.

Reply:
@amandahla would you mind opening that issue please?

Reply:
I found one already open:
https://github.com/canonical/operator/issues/762